File: Week_03/java/src/combinations/_77_Combinations.java
package combinations;

import java.util.ArrayList;
import java.util.List;

/**
 * 第一遍：2020/09/08周二 ✅
 * 第二遍：2020/06/08周一
 * 第三遍：2020/06/14周日
 * 第四遍：2020/06/28周日
 * todo: 理解多种解法
 */
public class _77_Combinations {
    static class UsingSequence {
        List<Integer> path = new ArrayList<>();
        List<List<Integer>> ret = new ArrayList<>();

        public List<List<Integer>> combine(int n, int k) {
            dfs(1, n, k);
            return ret;
        }

        void dfs(int i, int n, int k) {
            if (path.size() + n + 1 - i < k) return;
            if (path.size() == k) {
                ret.add(new ArrayList<>(path));
                return;
            }
            path.add(i);
            dfs(i + 1, n, k);
            path.remove(path.size() - 1);
            dfs(i + 1, n, k);
        }
    }
}

File: Week_03/kotlin/src/combinations/_77_Combinations.kt
package combinations

class _347_UsingSequence {
    fun combine(n: Int, k: Int): List<List<Int>> {
        val path = mutableListOf<Int>()
        val ret = mutableListOf<List<Int>>()
        fun dfs(i: Int) {
            if (path.size + n - i + 1 < k) return
            if (path.size == k) {
                ret += path.toList()
                return
            }
            path += i
            dfs(i + 1)
            path.removeAt(path.lastIndex)
            dfs(i + 1)
        }
        dfs(1)
        return ret
    }
}

class _77_奇技淫巧 {
    fun combine(n: Int, k: Int): List<List<Int>> {
        val ret = mutableListOf<List<Int>>()
        val path = IntArray(k)
        var i = 0
        while (i >= 0) {
            path[i]++
            when {
                path[i] > n -> i--
                i == k - 1 -> ret += path.toList()
                else -> {
                    i++
                    path[i] = path[i - 1]
                }
            }
        }
        return ret
    }
}